node_modules/col-activities/lib/notifications/daily.js

```diff
@@ -0,0 +1,548 @@
+ * @param  {Function}       callback        Standard callback function
+ * @api private
+ */
+var collectCourse = module.exports.collectCourse = function(course, callback) {
+  var start = Date.now();
+
+  // 1. Get all the data for this course
+  getCourseData(course, function(err, data) {
+    if (err) {
+      // Don't pass the error up
+      return callback();
+    }
+
+    // Return early if there's no course data
+    if (_.isEmpty(data.assets) && _.isEmpty(data.whiteboards)) {
+      return callback();
+    }
+
+    // 2. Iterate through all the users in this course and notify them if necessary
+    var userIterator = handleUser.bind(null, course, data);
+    async.eachSeries(data.users, userIterator, function() {
+      log.info('Sent emails for course %d, took %d ms', course.id, Date.now() - start);
+      return callback();
+    });
+  });
+};
+
+/**
+ * Send notifications for a user in a course.
+ *
+ * Note that this function won't pass up any errors.
+ *
+ * @param  {Course}     course        The course for which to send out the daily notifications
+ * @param  {Object}     data          The relevant data in the given course
+ * @param  {User}       user          The user to notify
+ * @param  {Function}   callback      Standard callback function
+ * @api private
+ */
+var handleUser = function(course, data, user, callback) {
+  // Ignore users that don't have an email address
+  if (!user.canvas_email) {
+    return callback();
+  }
+
+  // Get all the activities for the user
+  var activities = getActivitiesForUser(data, user);
+
+  // If nothing noteworthy happened for this user, then they should not get an email
+  if (_.isEmpty(activities)) {
+    return callback();
+  }
+
+  var subject = getSubject(course, activities, user);
+
+  // The data that's used in the daily email template
+  var data = {
+    'activities': activities,
+    'getSummaryActors': getSummaryActors,
+  };
+  EmailUtil.sendEmail(subject, user, course, data, 'daily', function(err) {
+    // Don't pass errors up as otherwise all the following users would be ignored.
+    // The error itself is logged in the sendEmail implementation
+    return callback();
+  });
+};
+
+/**
+ * Get the subject for the daily email
+ *
+ * @param  {Course}                                 course        The course for which to send out the daily notifications
+ * @param  {(AssetActivity|WhiteboardActivity)[]}   activities    The set of activities for which to generate an email subject
+ * @param  {User}                                   currentUser   The user to whom the email will be sent
+ * @return {String}                                               An appropriate email subject
+ * @api private
+ */
+var getSubject = function(course, activities, currentUser) {
+  var subject = null;
+  if (activities.length === 1) {
+    var activity = activities[0];
+    if (activity.type === 'asset_comment') {
+      var commenters = _.pluck(activity.asset.comments, 'user');
```

Comment:
Note that extracting the `commenters` and `summaryActors` is the same in this block and the block below